amaranth/lib/crc/__init__.py

```diff
@@ -0,0 +1,417 @@
+
+
+class Processor(Elaboratable):
+    """
+    Cyclic redundancy check (CRC) processor module.
+
+    This module generates CRCs from an input data stream, which can be used
+    to validate an existing CRC or generate a new CRC. It is configured by
+    the :py:class:`Parameters` class, which can handle most forms of CRCs.
+    Refer to that class's documentation for a description of the parameters.
+
+    The CRC value is updated on any clock cycle where ``valid`` is asserted,
+    with the updated value available on the ``crc`` output on the subsequent
+    clock cycle. The latency is therefore one clock cycle, and the throughput
+    is one data word per clock cycle.
+
+    The CRC is reset to its initial value whenever ``start`` is asserted.
+    ``start`` and ``valid`` may be asserted on the same clock cycle, in which
+    case a new CRC computation is started with the current value of ``data``.
+
+    With ``data_width=1``, a classic bit-serial CRC is implemented for the
+    given polynomial in a Galois-type shift register. For larger values of
+    ``data_width``, a similar architecture computes every new bit of the
+    CRC in parallel.
+
+    The ``match_detected`` output may be used to validate data with a trailing
+    CRC (also known as a codeword). If the most recently processed word(s) form
+    the valid CRC of all the previous data since ``start`` was asserted, the
+    CRC register will always take on a fixed value known as the residue.  The
+    ``match_detected`` output indicates whether the CRC register currently
+    contains this residue.
+
+    Parameters
+    ----------
+    parameters : Parameters
+        CRC parameters.
+
+    Attributes
+    ----------
+    start : Signal(), in
+        Assert to indicate the start of a CRC computation, re-initialising
+        the CRC register to the initial value. May be asserted simultaneously
+        with ``valid`` or by itself.
+    data : Signal(data_width), in
+        Data word to add to CRC when ``valid`` is asserted.
+    valid : Signal(), in
+        Assert when ``data`` is valid to add the data word to the CRC.
+    crc : Signal(crc_width), out
+        Registered CRC output value, updated one clock cycle after ``valid``
+        becomes asserted.
+    match_detected : Signal(), out
+        Asserted if the current CRC value indicates a valid codeword has been
+        received.
+    """
+    def __init__(self, parameters):
+        if not isinstance(parameters, Parameters):
+            raise TypeError("Parameters must be of type Parameters")
+        self.crc_width = parameters.crc_width
+        self.data_width = parameters.data_width
+        self.polynomial = parameters.polynomial
+        self.initial_crc = Const(parameters.initial_crc, self.crc_width)
+        self.reflect_input = parameters.reflect_input
+        self.reflect_output = parameters.reflect_output
+        self.xor_output = parameters.xor_output
```

Comment:
I don't think it's right to have public writable undocumented fields with these parameters--we could have a single read-only `parameters` property, or not have any public properties at all. (I have a weak preference for the latter, but no objection to the former.)

Reply:
Sure, I've changed them to private attributes for Parameters and Processor. I've left the ones on `Algorithm` unchanged though, is that OK? I feel like an argument could be made for documenting and making them public on Parameters too since it's just the ones from algorithm and the data width, but I don't think anyone's likely to need to use them -- you can just create/modify Algorithm and then the Parameters instance is quite transient on the way to Processor or computation.